services/user/CommonSys/common/packages/wasm-loader/src/main.ts

```diff
@@ -17,68 +33,242 @@ document.querySelector<HTMLDivElement>("#app")!.innerHTML = `
   </div>
 `;
 
+const supervisorDomain = siblingUrl(null, "supervisor-sys");
+
 interface Importables {
     [key: string]: string;
 }
 
+export interface Action {
+    service: string;
+    action: string;
+    args: string;
+}
+let addedActions: Action[];
+
 const wasmUrlToIntArray = (url: string) =>
     fetch(url)
-        .then((res) => res.arrayBuffer())
+        .then((res) => {
+            if (!res.ok) throw new DownloadFailed(url);
+            return res.arrayBuffer();
+        })
         .then((buffer) => new Uint8Array(buffer));
 
-const onPluginCallRequest = async ({
-    id,
-    args,
-    precomputedResults
-}: PluginCallPayload) => {
-    const [pluginBytes, parserBytes, wasmComponent] = await Promise.all([
-        wasmUrlToIntArray("/plugin.wasm"),
-        wasmUrlToIntArray("/common/component_parser.wasm"),
-        import("rollup-plugin-wit-component")
-    ]);
-
-    const { provider } = await wasmComponent.load(parserBytes);
-    const ComponentParser = new provider.ComponentParser();
-    const { wit } = ComponentParser.parse("component", pluginBytes);
-    const parsedFunctions = parseFunctions(wit);
-
-    const serviceImports = serviceMethodIndexToImportables(
-        parsedFunctions,
-        args.service,
-        id,
-        precomputedResults.map((res) => ({
-            method: res.method,
-            result: res.result
-        }))
-    );
+const parserCacheKey: string = "component-parser";
+const componentParser = async () => {
+    if (!cache.exists(parserCacheKey)) {
+        let parserBytes: Uint8Array;
+        try {
+            parserBytes = await wasmUrlToIntArray(
+                "/common/component_parser.wasm"
+            );
+        } catch (e) {
+            console.error("Loader: component_parser download failed");
+            throw new ParserDownloadFailed();
+        }
+        // Performance improvement: Target wasm32-unknown-unknown when compiling
+        //   parser instead of wasm32-wasi, then loading can be simplified without wasi
+        //   shims.
+        cache.cacheData(parserCacheKey, await load(parserBytes));
+        return cache.getCachedData(parserCacheKey);
+    } else return cache.getCachedData(parserCacheKey);
+};
```

Comment:
Can we flip this around to avoid more nesting?

```js
if (cache.exists(parserCacheKey)) {
    return cache.getCachedData(parserCacheKey);
}

let parserBytes: Uint8Array;
...
```

Reply:
Same for the next couple functions.

Reply:
I guess I don't need to think about tail call optimization in js

Reply:
Fixed